plugins/experimental/traffic_dump/traffic_dump.cc

```diff
@@ -315,12 +332,175 @@ session_aio_handler(TSCont contp, TSEvent event, void *edata)
     return TS_SUCCESS;
   }
   default:
-    TSDebug(PLUGIN_NAME, "session_aio_handler(): unhandled events %d", event);
+    TSDebug(PLUGIN_NAME, "%s: unhandled events %d", __func__, event);
     return TS_ERROR;
   }
   return TS_SUCCESS;
 }
 
+std::string
+request_body_get(TSHttpTxn txnp)
+{
+  TSDebug(PLUGIN_NAME, "getting the body");
+  TSIOBufferReader post_buffer_reader = TSHttpTxnPostBufferReaderGet(txnp);
+  int64_t read_avail                  = TSIOBufferReaderAvail(post_buffer_reader);
+  if (read_avail == 0) {
+    TSIOBufferReaderFree(post_buffer_reader);
+    return "";
+  }
+
+  // std::string has no reserving allocator. So create an explicitly empty
+  // string then reserve the bytes for later population.
+  std::string body_string;
+  body_string.reserve(read_avail + 1); // '+ 1' for null termination.
+
+  int64_t consumed      = 0;
+  int64_t data_len      = 0;
+  const char *char_data = NULL;
+  TSIOBufferBlock block = TSIOBufferReaderStart(post_buffer_reader);
+  while (block != NULL) {
+    char_data = TSIOBufferBlockReadStart(block, post_buffer_reader, &data_len);
+    body_string.replace(consumed, data_len, char_data, data_len);
+    consumed += data_len;
+    block = TSIOBufferBlockNext(block);
+  }
+  body_string[consumed] = '\0';
+  TSIOBufferReaderFree(post_buffer_reader);
+
+  TSDebug(PLUGIN_NAME, "returning %ld body bytes", body_string.size());
+  return body_string;
+}
+
+/** Print the json data for a transaction
+ *
+ * @param[in] txnp The reference to the transaction.
+ * @param[in] ssnData The plugin's data for the session.
+ * @param[in] txnData The plugin's data for the transaction.
+ */
+void
+print_transaction_content(TSHttpTxn txnp, SsnData *ssnData, TxnData *txnData)
+{
+  // Get UUID
+  char uuid[TS_CRUUID_STRING_LEN + 1];
+  TSAssert(TS_SUCCESS == TSClientRequestUuidGet(txnp, uuid));
+
+  // Generate per transaction json records
+  std::string txn_info;
```

Comment:
I recommend doing a `reserve` here of say 4K, to cut down on allocation calls in the  common case.

Reply:
Sounds good.